docs/rfcs/20220207_connections_per_ip.md

```diff
@@ -0,0 +1,160 @@
+* Prevent rogue clients from consuming unbounded memory (we allocate
+  some userspace buffer space to all connections, and the kernel
+  allocates some buffers too).
+* Help users notice if a client is acting strangely: an application
+  author opening large numbers of connections will tend to notice
+  themselves when they hit the limit, rather than the operator
+  of the cluster noticing when the cluster performance is impacted.
+* Provide a client allow-listing/deny-listing mechanism (e.g.
+  kafka_max.connections_per_ip is set to zero and
+  kafka_max_connections_per_ip_overrides specifies
+  permitted clients)
+* Continuity of experience for Kafka users accustomed to
+  max.connections.per.ip
+
+## How
+
+### Accepting a connection
+
+If both configuration properties are unset, then no action is taken: this
+preserves the existing unlimited behaviour.
+
+A sharded service storing per-client tokens for open connections, clients are
+mapped to shards by hash of their IP.
+
+In kafka::protocol::apply, acquire a token at start and relinquish it at end,
+probably using a RAII structure, although this function already provides a
+convenient localized place to take+release a resource explicitly.
+
+On first connection from a client, the shard servicing the connection makes a
+cross-shard call to the core that holds state for this client IP hash.  This
+initializes the client's state:
+
+struct client_conn_quota { // The max per connection.  This is set by applying
+// the combination of per_ip and per_ip_override settings.  uint32_t total;
+
+    // How many connections are currently open.  This may be at // most
+`total`, although it may exceed it transiently if // configuration properties
+are decreased during operation.  uint32_t in_use; }
+
+The connection quota state is stored in a absl::btree_map: since the overall
+number of unique clients maybe high (so a flat store is risky), but the entries
+are small (so a std::map would generate a lot of tiny allocations).
+
+### Optimization: caching tokens
+
+In the case where the client is not close to exceeding its limits, we can avoid
+the overhead of a cross-core call by caching some per-client tokens on each
+shard.
+
+This cache can be pre-primed on the first client connection, if the total
+tokens for the client is greater than the core count: that way, subsequent
+connections from the same client will usually be accepted without the need for
+a cross-core call.
+
+### Handling a configuration change
+
+Changes to the configuration properties affect all outstanding
+client_conn_quota objects: the `total` attribute must be recalculated.
+
+If `total` is decreased, then it is also necessary to call out to all other
+shards and revoke any cached tokens that are held in excess of the new limit.
+
+## Impact
+
+When this feature is enabled, a per-connection latency overhead is added, as
+the shard handling the connection must dispatch a call to the shard that owns
+the client IP state.
```

Comment:
Is this latency in the order of microseconds? If so, it might be worth mentioning because I'd see "microseconds" and think "ok so I don't really need to care". I'd expect the user to be more impacted by the connection congestion window, and even moreso sasl / api version negotiation.

Reply:
Yes, this is a single-digit microseconds (or even nanos) thing.  I call it out not because I think anyone is necessarily going to notice (it's likely to be faster than the TCP handshake), but to record that this feature isn't a free lunch (as we add more and more features like this, there will be a cumulative effect)